avro-builder/src/main/java/com/linkedin/avroutil1/builder/operations/codegen/own/AvroUtilCodeGenOp.java

```diff
@@ -51,10 +69,76 @@ public void run(OperationContext opContext) throws Exception {
+    AvroParseContext context = new AvroParseContext();
+    AvscParser parser = new AvscParser();
+    List<AvscParseResult> parsedFiles = new ArrayList<>();
 
-    generator.generateCode();
+    for (Path p : avscFiles) {
+      AvscParseResult fileParseResult = parser.parse(p);
+      Throwable parseError = fileParseResult.getParseError();
+      if (parseError != null) {
+        throw new IllegalArgumentException("failed to parse file " + p.toAbsolutePath(), parseError);
+      }
+      context.add(fileParseResult);
+      parsedFiles.add(fileParseResult);
+    }
+
+    //resolve any references across files that are part of this op (anything left would be external)
+    context.resolveReferences();
+
+    if (context.hasExternalReferences()) {
+      //TODO - better formatting
+      throw new UnsupportedOperationException(
+          "unresolved referenced to external schemas: " + context.getExternalReferences());
+    }
+
+    SpecificRecordClassGenerator generator = new SpecificRecordClassGenerator();
+    List<JavaFileObject> specificRecords = new ArrayList<>();
+    for (AvscParseResult parsedFile : parsedFiles) {
+      JavaFileObject javaFileObject =
+          generator.generateSpecificRecordClass((AvroNamedSchema) parsedFile.getTopLevelSchema(),
+              SpecificRecordGenerationConfig.BROAD_COMPATIBILITY);
+      specificRecords.add(javaFileObject);
+    }
+
+    writeJavaFilesToDisk(specificRecords, config.getOutputSpecificRecordClassesRoot());
+
+    //TODO - look for dups
+
+  }
+
+  private void writeJavaFilesToDisk(Collection<JavaFileObject> javaClassFiles, File outputFolder) {
+    //make sure the output folder exists
+    if (!outputFolder.exists() && !outputFolder.mkdirs()) {
+      throw new IllegalStateException("unable to create output folder " + outputFolder);
+    }
+
+    //write out the files we generated
+    for (JavaFileObject javaClass : javaClassFiles) {
+      File outputFile = new File(outputFolder, javaClass.getName());
+
+      if (outputFile.exists()) {
+        //TODO - make this behaviour configurable (overwite, ignore, ignore_if_identical, etc)
+        throw new IllegalStateException("output file " + outputFile + "already exists");
+      } else {
+        File parentFolder = outputFile.getParentFile();
+        if (!parentFolder.exists() && !parentFolder.mkdirs()) {
+          throw new IllegalStateException("unable to create output folder " + outputFolder);
+        }
+      }
+
+      try (FileOutputStream fos = new FileOutputStream(outputFile, false);
+          OutputStreamWriter writer = new OutputStreamWriter(fos, StandardCharsets.UTF_8)) {
+        Reader reader = javaClass.openReader(true);
+        IOUtils.copy(reader, writer);
+        reader.close();
+
+        writer.flush();
+        fos.flush();
+      } catch (Exception e) {
+        throw new IllegalStateException("while writing file " + outputFile, e);
+      }
```

Comment:
Let me know if there's a cleaner way to handle this Reader -> Writer copy.

Reply:
nit - you might be able to move the reader initialization into the try(), which would save you the need to close() it.